modules/transactions/src/main/java/org/apache/ignite/internal/tx/impl/TxManagerImpl.java

```diff
@@ -854,4 +891,66 @@ static boolean isRecoverable(Throwable throwable) {
             return false;
         }
     }
+
+    /**
+     * Handles messages of {@link TxMessageGroup}.
+     *
+     * @see NetworkMessageHandler#onReceived(NetworkMessage, String, Long)
+     */
+    private void onReceiveTxNetworkMessage(NetworkMessage message, String senderConsistentId, @Nullable Long correlationId) {
+        if (!(message instanceof RwTransactionsFinishedRequest)) {
+            return;
+        }
+
+        assert correlationId != null : senderConsistentId;
+
+        int targetCatalogVersion = ((RwTransactionsFinishedRequest) message).targetSchemaVersion();
+
+        boolean finished = isRwTransactionsFinished(targetCatalogVersion);
+
+        clusterService.messagingService().respond(senderConsistentId, toRwTransactionsFinishedResponse(finished), correlationId);
+    }
+
+    /**
+     * Returns the transaction catalog version.
+     *
+     * @param beginTs Transaction begin timestamp.
+     */
+    private int txCatalogVersion(HybridTimestamp beginTs) {
+        return catalogService.activeCatalogVersion(beginTs.longValue());
+    }
+
+    /**
+     * Returns the transaction catalog version.
+     *
+     * @param txId Transaction ID.
+     */
+    private int txCatalogVersion(UUID txId) {
+        return txCatalogVersion(TransactionIds.beginTimestamp(txId));
+    }
+
+    /**
+     * Returns {@code true} iff the requested catalog version is registered locally and all RW transactions before it have completed,
+     * otherwise {@code false}.
+     *
+     * @param catalogVersion Catalog version to check.
+     */
+    boolean isRwTransactionsFinished(int catalogVersion) {
+        if (catalogVersion > catalogService.latestCatalogVersion()) {
+            // Requested catalog version has not yet registered locally.
```

Comment:
Another problem is that this check is not atomic wrt the increment, so it might happen that we respond with 'we don't have transactions started on versions before bersion 10' and then start a transaction on version 9 (or even an earlier version).

I think this can be resolved if we add a new field for tracking the most recent **activated** catalog version we know about (`lastKnownActiveCatalogVer`). We will update it here, just before doing the check. (The update can be done under a write lock). When doing an increment, we would do a loop:

1. beginTs = now
2. catalogVer  = catalogVerByTs(beginTs)
3. [under read lock] if catalogVer >= lastKnownActiveCatalogVer then do increment; else goto 1

Reply:
When working with a catalog, we always deal with active versions of the catalog; when we get the latest version, we get the latest active version.

We will have an additional mechanism for each partition, which will allow us to wait for all operations until the version in which the index appears, I think we can cover corner cases with it.

Otherwise, we will need to do a blocking algorithm, which we would like to avoid.